include/nbl/video/IGPUCommandPool.h

```diff
@@ -334,103 +234,119 @@ class NBL_API IGPUCommandPool : public core::IReferenceCounted, public IBackendO
             {
-                listTail->setNext(segment);
-                listTail = listTail->getNext();
-            }
-            else
-            {
-                assert(!listHeadItr.cmd && !listHeadItr.segment && "List should've been empty.");
-
-                listTail = segment;
-                listHeadItr = segment->begin();
-            }
+                auto cmdMem = list.tail->allocate<Cmd>(args...);
+                if (!cmdMem)
+                    return nullptr;
 
-            return bool(segment);
-        }
+                return new (cmdMem) Cmd(std::forward<Args>(args)...);
+            };
 
-        void deleteList(CCommandSegment::Iterator& headItr, CCommandSegment*& tail)
-        {
-            if (tail)
+            auto cmd = newCmd();
+            if (!cmd)
             {
-                assert(headItr.segment && headItr.cmd);
+                if (!appendToList(list))
+                    return nullptr;
 
-                auto freeSegment = [this](CCommandSegment* segment)
+                cmd = newCmd();
+                if (!cmd)
                 {
-                    if (segment)
-                    {
-                        segment->~CCommandSegment();
-                        m_pool.deallocate(segment, COMMAND_SEGMENT_SIZE);
-                    }
-                };
-
-                CCommandSegment* prevSegment = nullptr;
-                for (auto& itr = headItr; itr != tail->end();)
-                {
-                    prevSegment = itr.segment;
-                    auto prevCmd = *(itr++);
+                    assert(false);
+                    return nullptr;
+                }
+            }
 
-                    prevCmd->~ICommand();
+            return cmd;
+        }
 
-                    // No need to deallocate prevCmd because:
-                    // - it has been allocated from the LinearAddressAllocator where deallocate is a No-OP
-                    // - the memory will get reclaimed in `~LinearAddressAllocator` when whole segment is destroyed
+        // Nullifying the head of the passed segment list is NOT the responsibility of deleteList.
+        inline void deleteList(CCommandSegment* head)
+        {
+            if (!head)
+                return;
 
-                    if (itr.segment == prevSegment)
-                        continue;
+            if (head == m_head)
+                m_head = head->getNextHead();
 
-                    // Segment has changed, we need to free the previous one.
-                    freeSegment(prevSegment);
-                }
+            auto oneBefore = head->getPrevHead();
+            auto oneAfter = head->getNextHead();
+            CCommandSegment::linkHeads(oneBefore, oneAfter);
 
-                // The last segment is always freed outside the loop.
-                assert(tail == prevSegment);
-                freeSegment(tail);
+            if (!oneBefore && !oneAfter)
+                m_head = nullptr;
```

Comment:
you don't need this the assignement at the very start of the function takes care of this and more

(when you call `getNextHead()` on the last item, it returns nullptr)

Reply:
Resolved.